File: 97-interleaving-string/interleaving-string.cpp
class Solution {
public:
    unordered_map<string, bool> cache;

    bool dfs(int i, int j, string s1, string s2, string s3){
        if (i == s1.length() && j == s2.length()) return true;
        string key = to_string(i) + "_" + to_string(j);
        if (cache.contains(key)) return cache[key];
        if (i < s1.length() && s1[i] == s3[i+j] && dfs(i+1, j, s1, s2, s3)) return true;
        if (j < s2.length() && s2[j] == s3[i+j] && dfs(i, j+1, s1, s2, s3)) return true;
        cache[key] = false;
        return cache[key];
    } 

    bool isInterleave(string s1, string s2, string s3) {
        if (s1.length() + s2.length() != s3.length()) return false;
        return dfs(0, 0, s1, s2, s3);
    }
};

File: 97-interleaving-string/interleaving-string.java
class Solution {
    
    HashMap<String, Boolean> cache = new HashMap<String, Boolean>();

    public boolean dfs(int i, int j, String s1, String s2, String s3){
        if (i == s1.length() && j == s2.length()) return true;
        String key = String.valueOf(i) + "_" + String.valueOf(j);
        if (cache.containsKey(key)) return cache.get(key);
        if (i < s1.length() && s1.charAt(i) == s3.charAt(i + j) && dfs(i+1, j, s1, s2, s3)) return true;
        if (j < s2.length() && s2.charAt(j) == s3.charAt(i + j) && dfs(i, j+1, s1, s2, s3)) return true;
        cache.put(key,false);
        return cache.get(key);
    } 


    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) return false;
        return dfs(0, 0, s1, s2, s3);
    }
}